src/Avalonia.Base/Styling/Setter.cs

```diff
@@ -57,16 +57,72 @@ public Setter(AvaloniaProperty property, object value)
             }
         }
 
-        public ISetterInstance Instance(IStyleable target)
-        {
-            target = target ?? throw new ArgumentNullException(nameof(target));
+        bool IValueEntry.HasValue => true;
+        AvaloniaProperty IValueEntry.Property => EnsureProperty();
+
+        public override string ToString() => $"Setter: {Property} = {Value}";
 
+        void IValueEntry.Unsubscribe() { }
+
+        ISetterInstance ISetter.Instance(IStyleInstance instance, IStyleable target)
+        {
+            if (target is not AvaloniaObject ao)
+                throw new InvalidOperationException("Don't know how to instance a style on this type.");
             if (Property is null)
-            {
                 throw new InvalidOperationException("Setter.Property must be set.");
+            if (Property.IsDirect && instance.HasActivator)
+                throw new InvalidOperationException(
+                    $"Cannot set direct property '{Property}' in '{instance.Source}' because the style has an activator.");
+
+            if (Value is IBinding binding)
```

Comment:
Switch statement here would be easier for the eyes :)

Reply:
Yeah I did actually try out a switch statement here but it didn't look as clean to me:

```
            return Value switch
            {
                IBinding binding => SetBinding((StyleInstance)instance, ao, binding),
                ITemplate template when !typeof(ITemplate).IsAssignableFrom(Property.PropertyType) =>
                    new PropertySetterTemplateInstance(Property, template),
                _ when !Property.IsValidValue(Value) =>
                    throw new InvalidCastException($"Setter value '{Value}' is not a valid value for property '{Property}'."),
                _ when Property.IsDirect => SetDirectValue(target),
                _ => this,
            };
```

Same number of lines, but reads less clear than the `if` to me, mainly because it's testing a few different things beyond just `Value`.